File: android/src/main/java/com/irfan/xendit_flutter/XenditFlutterPlugin.java
package com.irfan.xendit_flutter;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.xendit.AuthenticationCallback;
import com.xendit.Models.Address;
import com.xendit.Models.AuthenticatedToken;
import com.xendit.Models.Authentication;
import com.xendit.Models.BillingDetails;
import com.xendit.Models.Card;
import com.xendit.Models.CardInfo;
import com.xendit.Models.Customer;
import com.xendit.Models.Token;
import com.xendit.Models.XenditError;
import com.xendit.TokenCallback;
import com.xendit.Xendit;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;


/** XenditFlutterPlugin */
public class XenditFlutterPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private Xendit xendit;
  private Activity activity;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "plugins.flutter.io/xendit");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method){
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
        break;
      case "createToken":
        createToken(call, result);
        break;
      case "createAuthentication":
        createAuthentication(call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
//    if (call.method.equals("getPlatformVersion")) {
//      result.success("Android " + android.os.Build.VERSION.RELEASE);
//    } else {
//      result.notImplemented();
//    }
  }

  private void init(String publishedKey){
    if(activity != null){
      xendit = new Xendit(activity.getApplication().getApplicationContext(), publishedKey, activity);
    }
  }

  private void createToken(@NonNull MethodCall call, @NonNull final Result result){
    if(call.hasArgument("publishedKey")){
      init(call.argument("publishedKey"));
    }

    Integer amount = call.hasArgument("amount") ? call.argument("amount") : 0;
    String onBehalfOf = call.hasArgument("onBehalfOf") ? call.argument("onBehalfOf") : "";
    Boolean shouldAuthenticate = call.hasArgument("shouldAuthenticate") ? call.argument("shouldAuthenticate")
            : true;
    boolean isMultipleUse = call.hasArgument("isMultipleUse") ? call.argument("isMultipleUse") : false;

    Card card = cardFrom(call);
    BillingDetails billingDetails = billingFrom(call);
    Customer customer = customerFrom(call);
    String currency = call.hasArgument("currency") ? call.argument("currency") : "IDR";

    TokenCallback tokenCallback = new TokenCallback() {
      @Override
      public void onSuccess(Token token) {
        result.success(tokenToMap(token));
      }

      @Override
      public void onError(XenditError xenditError) {
        result.error(xenditError.getErrorCode(), xenditError.getErrorMessage(), "");
      }
    };
//    xendit.createSingleUseToken(card, amount, true, "", billingDetails, customer, "IDR", new TokenCallback() {
//      @Override
//      public void onSuccess(Token token) {
//        // Handle successful tokenization
//        // System.out.println("Token ID: " + token.getId());
//        // return tokenToMap(token);
//        result.success(tokenToMap(token));
//      }
//
//      @Override
//      public void onError(XenditError xenditError) {
//        // Handle error
//        Map<String, Object> myMap = new HashMap<>();
//        myMap.put("key1", 3.14);
//        myMap.put("key2", 2.71);
//        // result.success(myMap);
//        result.error(xenditError.getErrorCode(), xenditError.getErrorMessage(), myMap);
//      }
//    });
    if(isMultipleUse){
      xendit.createMultipleUseToken(card, onBehalfOf, billingDetails,customer, tokenCallback);
    }else{
      xendit.createSingleUseToken(card, amount.toString(), shouldAuthenticate, onBehalfOf, billingDetails, customer, currency, tokenCallback);
    }
  }


  private void createAuthentication(@NonNull MethodCall call, @NonNull final Result result) {
    if (call.hasArgument("publishedKey")) {
      init(call.argument("publishedKey"));
    }

    String tokenId = call.argument("tokenId");
    Integer _amount = call.argument("amount");
    int amount = _amount != null ? _amount : 0;
    String currency = null;

    if (call.hasArgument("currency")) {
      currency = call.argument("currency");
    }

    xendit.createAuthentication(tokenId, amount, currency, new AuthenticationCallback() {
      @Override
      public void onSuccess(Authentication authentication) {
        result.success(authenticationToMap(authentication));
      }

      @Override
      public void onError(XenditError error) {
        result.error(error.getErrorCode(), error.getErrorMessage(), "");
      }
    });
  }

  private Card cardFrom(@NonNull MethodCall call) {
    if (call.hasArgument("card")) {
      Map<String, Object> map = call.argument("card");

      if (map != null) {
        String creditCardNumber = (String) map.get("creditCardNumber");
        String creditCardCVN = (String) map.get("creditCardCVN");
        String expirationMonth = (String) map.get("expirationMonth");
        String expirationYear = (String) map.get("expirationYear");
        return new Card(creditCardNumber, expirationMonth, expirationYear, creditCardCVN);
      }
    }

    return null;
  }

  private BillingDetails billingFrom(@NonNull MethodCall call) {
    if (call.hasArgument("billingDetails")) {
      BillingDetails billingDetails = new BillingDetails();
      Map<String, Object> map = call.argument("billingDetails");

      if (map != null) {
        billingDetails.setGivenNames((String) map.get("givenNames"));
        billingDetails.setSurname((String) map.get("surname"));
        billingDetails.setEmail((String) map.get("email"));
        billingDetails.setMobileNumber((String) map.get("mobileNumber"));
        billingDetails.setPhoneNumber((String) map.get("phoneNumber"));

        Map aMap = (Map) map.get("address");

        Address address = new Address();
        if (aMap != null) {
          address.setCountry((String) aMap.get("country"));
          address.setStreetLine1((String) aMap.get("streetLine1"));
          address.setStreetLine2((String) aMap.get("streetLine2"));
          address.setCity((String) aMap.get("city"));
          address.setProvinceState((String) aMap.get("provinceState"));
          address.setPostalCode((String) aMap.get("postalCode"));
          address.setCategory((String) aMap.get("category"));
          billingDetails.setAddress(address);
        } else {
          billingDetails.setAddress(address);
        }
      }
      return billingDetails;
    }
    return new BillingDetails();
  }

  private Customer customerFrom(@NonNull MethodCall call) {
    Customer customer = new Customer();
    if (call.hasArgument("customer")) {
      Map<String, Object> map = call.argument("customer");

      if (map != null) {
        customer.setReferenceId((String) map.get("referenceId"));
        customer.setEmail((String) map.get("email"));
        customer.setGivenNames((String) map.get("givenNames"));
        customer.setSurname((String) map.get("surname"));
        customer.setDescription((String) map.get("description"));
        customer.setMobileNumber((String) map.get("mobileNumber"));
        customer.setPhoneNumber((String) map.get("phoneNumber"));
        customer.setNationality((String) map.get("nationality"));
        customer.setDateOfBirth((String) map.get("dateOfBirth"));
        customer.setCardInfo((Map<String, String>) map.get("cardInfo"));

        if (map.containsKey("addresses")) {
          List aList = (List) map.get("addresses");

          if (aList != null && aList.size() > 0) {
            int count = aList.size();
            Address[] addresses = new Address[count];
            for (int i = 0; i < count; i++) {
              Map aMap = (Map) aList.get(i);

              if (aMap != null) {
                Address address = new Address();
                address.setCountry((String) aMap.get("country"));
                address.setStreetLine1((String) aMap.get("streetLine1"));
                address.setStreetLine2((String) aMap.get("streetLine2"));
                address.setCity((String) aMap.get("city"));
                address.setProvinceState((String) aMap.get("provinceState"));
                address.setPostalCode((String) aMap.get("postalCode"));
                address.setCategory((String) aMap.get("category"));
                addresses[i] = address;
              }
            }
            customer.setAddresses(addresses);
          }
        }
      }
      return customer;
    }
    return customer;
  }

  private Map<String, Object> authenticationToMap(Authentication authentication) {
    Map<String, Object> result = new HashMap<>();
    result.put("id", authentication.getId());
    result.put("creditCardTokenId", authentication.getCreditCardTokenId());
    result.put("payerAuthenticationUrl", authentication.getPayerAuthenticationUrl());
    result.put("status", authentication.getStatus());
    result.put("maskedCardNumber", authentication.getMaskedCardNumber());
    result.put("requestPayload", authentication.getRequestPayload());
    result.put("authenticationTransactionId", authentication.getAuthenticationTransactionId());
    result.put("cardInfo", cardInfoToMap(authentication.getCardInfo()));
    return result;
  }

  @SuppressWarnings("deprecation")
  private Map<String, Object> tokenToMap(Token token) {
    Map<String, Object> result = new HashMap<>();
    result.put("id", token.getId());
    result.put("status", token.getStatus());
    result.put("authenticationId", token.getAuthenticationId());
    result.put("maskedCardNumber", token.getMaskedCardNumber());
    result.put("should3ds", token.getShould_3DS());
    result.put("authentication", authenticatedTokenToMap(token.getAuthentication()));
    result.put("cardInfo", cardInfoToMap(token.getCardInfo()));
    return result;
  }

  private Map<String, Object> authenticatedTokenToMap(AuthenticatedToken authenticatedToken) {
    Map<String, Object> result = new HashMap<>();
    result.put("id", authenticatedToken.getId());
    result.put("status", authenticatedToken.getStatus());
    result.put("authenticationId", authenticatedToken.getAuthenticationId());
    result.put("payerAuthenticationUrl", authenticatedToken.getPayerAuthenticationUrl());
    result.put("maskedCardNumber", authenticatedToken.getMaskedCardNumber());
    result.put("jwt", authenticatedToken.getJwt());
    result.put("threedsVersion", authenticatedToken.getThreedsVersion());
    result.put("environment", authenticatedToken.getEnvironment());
    result.put("cardInfo", cardInfoToMap(authenticatedToken.getCardInfo()));
    return result;
  }

  private Map<String, Object> cardInfoToMap(CardInfo cardInfo) {
    Map<String, Object> result = new HashMap<>();
    result.put("bank", cardInfo.getBank());
    result.put("country", cardInfo.getCountry());
    result.put("type", cardInfo.getType());
    result.put("brand", cardInfo.getBrand());
    result.put("cardArtUrl", cardInfo.getCardArtUrl());
    result.put("fingerprint", cardInfo.getFingerprint());
    return result;
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    activity = binding.getActivity();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

  }

  @Override
  public void onDetachedFromActivity() {

  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}

File: ios/Classes/XenditFlutterPlugin.swift
import Flutter
import UIKit
import Xendit

public class XenditFlutterPlugin: NSObject, FlutterPlugin {
    
    private var flutterViewController: FlutterViewController?
    
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "plugins.flutter.io/xendit", binaryMessenger: registrar.messenger())
    let instance = XenditFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
          result("iOS " + UIDevice.current.systemVersion)
        case "createToken":
            if let window = UIApplication.shared.keyWindow?.rootViewController as? FlutterViewController{
                flutterViewController = window
            }else{
                print("saalah")
            }
            createToken(call: call,result: result)
        case "createAuthentication":
            if let window = UIApplication.shared.keyWindow?.rootViewController as? FlutterViewController{
                flutterViewController = window
            }else{
                print("saalah")
            }
            createAuthentication(call: call, result: result)
        default:
          result(FlutterMethodNotImplemented)
        }
      }
    
    public func createToken(call: FlutterMethodCall,result: @escaping FlutterResult){
        var argument:[String: Any] = [:]
        let cardDataArg = argument["card"]
        var publishableKey = ""
        if let arguments = call.arguments as? Dictionary<String, Any>,
           let _ = arguments["publishedKey"] as? String{
            publishableKey = arguments["publishedKey"] as! String
            argument = arguments
        }
        Xendit.publishableKey = publishableKey
        let cardData = cardFrom(call: call)
        
        let isMultipleUse = argument["isMultipleUse"] as? Bool ?? false
        let currency = argument["currency"] ?? "IDR"
        var amount: NSNumber? = nil;
        if argument["amount"] != nil{
            amount = NSNumber(value: (argument["amount"] as! Int64))
        }
        let onBehalfOf = argument["onBehalfOf"] ?? ""
        
        let customer = customerFrom(call: call)
        let billingDetails = billingFrom(call: call)

        
         let tokenizationRequest = XenditTokenizationRequest.init(cardData: cardData as! XenditCardData, isSingleUse: !isMultipleUse, shouldAuthenticate: true, amount: amount, currency: currency as! String)
         tokenizationRequest.customer = customer
         tokenizationRequest.billingDetails = billingDetails

         Xendit.createToken(fromViewController: flutterViewController!, tokenizationRequest: tokenizationRequest, onBehalfOf: onBehalfOf as! String) { [self] (token, error) in
             if let token = token {
                 result(tokenToMap(token: token))
             } else {
                 result(FlutterError(code: error?.errorCode ?? "500", message: error?.message, details: nil))
             }
         }
    }
  
    
    public func createAuthentication(call: FlutterMethodCall, result: @escaping FlutterResult){
        var argument:[String: Any] = [:]
        var publishableKey = ""
        if let arguments = call.arguments as? Dictionary<String, Any>,
           let _ = arguments["publishedKey"] as? String{
            publishableKey = arguments["publishedKey"] as! String
            argument = arguments
        }
        Xendit.publishableKey = publishableKey
        
        let tokenId = argument["tokenId"] as! String
        let amount = NSNumber(value: (argument["amount"] as! Int64))
        let currency = argument["currency"] as! String
        
        let xenditAuthReq = XenditAuthenticationRequest(tokenId: tokenId, amount: amount, currency: currency)
        Xendit.createAuthentication(fromViewController: flutterViewController!, authenticationRequest: xenditAuthReq, onBehalfOf: nil){ [self](authentication, error) in
            if authentication != nil {
                result(authToMap(authentication: authentication!))
            }else{
                result(FlutterError(code: error!.errorCode, message: error!.message, details: nil))
            }
        }
    }

    
    private func cardFrom(call: FlutterMethodCall) -> XenditCardData? {
        if let cardMap = call.arguments as? [String: Any],
           let cardArgs = cardMap["card"] as? [String:Any],
            
        
           let creditCardNumber = cardArgs["creditCardNumber"] as? String,
           let creditCardCVN = cardArgs["creditCardCVN"] as? String,
           let expirationMonth = cardArgs["expirationMonth"] as? String,
           let expirationYear = cardArgs["expirationYear"] as? String {
            var cardData =  XenditCardData.init(cardNumber: creditCardNumber, cardExpMonth: expirationMonth, cardExpYear: expirationYear)
            cardData.cardCvn = creditCardCVN
            return cardData
        }
        
        return nil
    }

    
    private func billingFrom(call: FlutterMethodCall) -> XenditBillingDetails {
        if let billingDetailsMap = call.arguments as? [String: Any],
           let billingDetailsArgs = billingDetailsMap["billingDetails"] as? [String: Any] {
            
            var billingDetails = XenditBillingDetails()
            
            billingDetails.givenNames = billingDetailsArgs["givenNames"] as? String ?? ""
            billingDetails.surname = billingDetailsArgs["surname"] as? String ?? ""
            billingDetails.email = billingDetailsArgs["email"] as? String ?? ""
            billingDetails.mobileNumber = billingDetailsArgs["mobileNumber"] as? String ?? ""
            billingDetails.phoneNumber = billingDetailsArgs["phoneNumber"] as? String ?? ""

            if let addressArgs = billingDetailsArgs["address"] as? [String: Any] {
                var address = XenditAddress()

                address.country = addressArgs["country"] as? String ?? ""
                address.streetLine1 = addressArgs["streetLine1"] as? String ?? ""
                address.streetLine2 = addressArgs["streetLine2"] as? String ?? ""
                address.city = addressArgs["city"] as? String ?? ""
                address.provinceState = addressArgs["provinceState"] as? String ?? ""
                address.postalCode = addressArgs["postalCode"] as? String ?? ""
                address.category = addressArgs["category"] as? String ?? ""

                billingDetails.address = address
            }

            return billingDetails
        }

        return XenditBillingDetails()
    }
    
    func customerFrom(call: FlutterMethodCall) -> XenditCustomer {
        let customer = XenditCustomer()

        if let customerMap = call.arguments as? [String: Any],
           let customerArgs = customerMap["customer"] as? [String: Any] {

            customer.referenceId = customerArgs["referenceId"] as? String
            customer.email = customerArgs["email"] as? String
            customer.givenNames = customerArgs["givenNames"] as? String
            customer.surname = customerArgs["surname"] as? String
            customer.description = customerArgs["description"] as? String
            customer.mobileNumber = customerArgs["mobileNumber"] as? String
            customer.phoneNumber = customerArgs["phoneNumber"] as? String
            customer.nationality = customerArgs["nationality"] as? String
            customer.dateOfBirth = customerArgs["dateOfBirth"] as? String
            // print(customer)
             


            if let addressesList = customerArgs["addresses"] as? [[String: Any]], !addressesList.isEmpty {
                var addresses = [XenditAddress]()

                for aMap in addressesList {
                    var address = XenditAddress()

                    address.country = aMap["country"] as? String
                    address.streetLine1 = aMap["streetLine1"] as? String
                    address.streetLine2 = aMap["streetLine2"] as? String
                    address.city = aMap["city"] as? String
                    address.provinceState = aMap["provinceState"] as? String
                    address.postalCode = aMap["postalCode"] as? String
                    address.category = aMap["category"] as? String

                    addresses.append(address)
                }

                customer.addresses = addresses
            }else{
                let address = XenditAddress()
                customer.addresses = [address]
            }
        }

        return customer
    }

    private func tokenToMap(token: XenditCCToken) -> [String: Any] {
        var result: [String: Any] = [:]
        result["id"] = token.id
        result["status"] = token.status
        result["authenticationId"] = token.authenticationId
        result["maskedCardNumber"] = token.maskedCardNumber
        result["authentication_url"] = (token.authenticationId)
        result["authentication_id"] = (token.authenticationId)
        result["cardInfo"] = cardInfoToMap(cardInfo: token.cardInfo)
        return result
    }
    
    private func authToMap(authentication: XenditAuthentication)-> [String:Any]{
        var result: [String: Any] = [:]
        result["id"] = authentication.id
        result["creditCardTokenId"] = authentication.tokenId
        result["payerAuthenticationUrl"] = authentication.authenticationURL
        result["status"] = authentication.status
        result["maskedCardNumber"] = authentication.maskedCardNumber
        result["requestPayload"] = authentication.requestPayload
        result["authenticationTransactionId"] = authentication.authenticationTransactionId
        result["cardInfo"] = cardInfoToMap(cardInfo: authentication.cardInfo)
        return result
    }

    private func cardInfoToMap(cardInfo: XenditCardMetadata?) -> [String: Any] {
        guard let cardInfo = cardInfo else {
            return [:]
        }

        var result: [String: Any] = [:]
        result["bank"] = cardInfo.bank
        result["country"] = cardInfo.country
        result["type"] = cardInfo.type
        result["brand"] = cardInfo.brand
        result["cardArtUrl"] = cardInfo.cardArtUrl
        result["fingerprint"] = cardInfo.fingerprint
        return result
    }

}
